File: backend/app/utils/payoff.py
"""Option strategy payoff calculator.

Computes P&L at expiry for a set of option/futures legs across a range of underlying prices.
"""

from dataclasses import dataclass


@dataclass
class Leg:
    strike: float
    instrument_type: str  # CE, PE, FUT
    transaction_type: str  # BUY, SELL
    quantity: int
    premium: float  # Premium paid/received per unit


def calculate_payoff(
    legs: list[Leg],
    spot_range: tuple[float, float] | None = None,
    num_points: int = 200,
) -> dict:
    """Calculate P&L across a range of underlying prices at expiry.

    Returns:
        {
            "points": [{"underlying_price": float, "pnl": float}, ...],
            "max_profit": float | None,
            "max_loss": float | None,
            "breakevens": [float, ...],
        }
    """
    if not legs:
        return {"points": [], "max_profit": None, "max_loss": None, "breakevens": []}

    # Determine price range if not provided
    strikes = [leg.strike for leg in legs if leg.instrument_type != "FUT"]
    if not strikes:
        strikes = [leg.strike for leg in legs]

    center = sum(strikes) / len(strikes) if strikes else 20000
    spread = max(strikes) - min(strikes) if len(strikes) > 1 else center * 0.1
    margin = max(spread * 2, center * 0.1)

    if spot_range is None:
        low = center - margin
        high = center + margin
    else:
        low, high = spot_range

    step = (high - low) / num_points

    points = []
    prev_pnl = None
    breakevens = []

    for i in range(num_points + 1):
        spot = low + i * step
        total_pnl = 0.0

        for leg in legs:
            multiplier = 1 if leg.transaction_type == "BUY" else -1

            if leg.instrument_type == "CE":
                intrinsic = max(0, spot - leg.strike)
                pnl = (intrinsic - leg.premium) * leg.quantity * multiplier
            elif leg.instrument_type == "PE":
                intrinsic = max(0, leg.strike - spot)
                pnl = (intrinsic - leg.premium) * leg.quantity * multiplier
            elif leg.instrument_type == "FUT":
                pnl = (spot - leg.strike) * leg.quantity * multiplier
            else:
                pnl = 0.0

            total_pnl += pnl

        points.append({"underlying_price": round(spot, 2), "pnl": round(total_pnl, 2)})

        # Detect breakeven (zero crossing)
        if prev_pnl is not None and prev_pnl * total_pnl < 0:
            # Linear interpolation
            prev_spot = low + (i - 1) * step
            ratio = abs(prev_pnl) / (abs(prev_pnl) + abs(total_pnl))
            be = prev_spot + ratio * step
            breakevens.append(round(be, 2))

        prev_pnl = total_pnl

    pnls = [p["pnl"] for p in points]
    max_profit = max(pnls) if pnls else None
    max_loss = min(pnls) if pnls else None

    # If max_profit or max_loss is at the boundary, it may be unbounded
    if pnls and pnls[-1] == max_profit and pnls[-1] > pnls[-2]:
        max_profit = None  # Unbounded profit
    if pnls and pnls[0] == max_loss and pnls[0] < pnls[1]:
        max_loss = None  # Unbounded loss

    return {
        "points": points,
        "max_profit": max_profit,
        "max_loss": max_loss,
        "breakevens": breakevens,
    }

File: frontend/src/lib/utils/payoff.ts
export interface PayoffLeg {
  strike: number;
  instrumentType: "CE" | "PE" | "FUT";
  transactionType: "BUY" | "SELL";
  quantity: number;
  premium: number;
}

export interface PayoffPoint {
  underlying_price: number;
  pnl: number;
}

export interface PayoffResult {
  points: PayoffPoint[];
  maxProfit: number | null;
  maxLoss: number | null;
  breakevens: number[];
}

export function calculatePayoff(
  legs: PayoffLeg[],
  numPoints: number = 200
): PayoffResult {
  if (legs.length === 0) {
    return { points: [], maxProfit: null, maxLoss: null, breakevens: [] };
  }

  const strikes = legs
    .filter((l) => l.instrumentType !== "FUT")
    .map((l) => l.strike);
  const allStrikes = strikes.length > 0 ? strikes : legs.map((l) => l.strike);

  const center =
    allStrikes.reduce((a, b) => a + b, 0) / allStrikes.length;
  const spread =
    allStrikes.length > 1
      ? Math.max(...allStrikes) - Math.min(...allStrikes)
      : center * 0.1;
  const margin = Math.max(spread * 2, center * 0.1);

  const low = center - margin;
  const high = center + margin;
  const step = (high - low) / numPoints;

  const points: PayoffPoint[] = [];
  let prevPnl: number | null = null;
  const breakevens: number[] = [];

  for (let i = 0; i <= numPoints; i++) {
    const spot = low + i * step;
    let totalPnl = 0;

    for (const leg of legs) {
      const multiplier = leg.transactionType === "BUY" ? 1 : -1;
      let pnl = 0;

      if (leg.instrumentType === "CE") {
        const intrinsic = Math.max(0, spot - leg.strike);
        pnl = (intrinsic - leg.premium) * leg.quantity * multiplier;
      } else if (leg.instrumentType === "PE") {
        const intrinsic = Math.max(0, leg.strike - spot);
        pnl = (intrinsic - leg.premium) * leg.quantity * multiplier;
      } else if (leg.instrumentType === "FUT") {
        pnl = (spot - leg.strike) * leg.quantity * multiplier;
      }

      totalPnl += pnl;
    }

    points.push({
      underlying_price: Math.round(spot * 100) / 100,
      pnl: Math.round(totalPnl * 100) / 100,
    });

    if (prevPnl !== null && prevPnl * totalPnl < 0) {
      const prevSpot = low + (i - 1) * step;
      const ratio = Math.abs(prevPnl) / (Math.abs(prevPnl) + Math.abs(totalPnl));
      breakevens.push(Math.round((prevSpot + ratio * step) * 100) / 100);
    }
    prevPnl = totalPnl;
  }

  const pnls = points.map((p) => p.pnl);
  let maxProfit: number | null = Math.max(...pnls);
  let maxLoss: number | null = Math.min(...pnls);

  if (pnls[pnls.length - 1] === maxProfit && pnls[pnls.length - 1] > pnls[pnls.length - 2]) {
    maxProfit = null;
  }
  if (pnls[0] === maxLoss && pnls[0] < pnls[1]) {
    maxLoss = null;
  }

  return { points, maxProfit, maxLoss, breakevens };
}
